Plugins Catalog: migrate state handling to Redux

leventebalogh opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
- [x] create a new reducer/actions/selectors/hooks for the plugin catalog
- [x] add API methods for fetching from multiple sources
- [x] migrate the Browse page to use the new reducer
- [x] migrate the Plugin details page to use the new reducer
- [x] make the Install/Uninstall functionality work with the new reducer
- [x] provide backwards-compatibility for parts of the app that were using the old plugins reducer
- [x] make sure the previous tests for Browse and Plugin details are working 
- [x] Fix - the plugin details get removed if the plugin-details page is loaded directly and then the user navigates to the browse
- [x] Fix - show the first tab when an app plugin gets uninstalled and a user was on a config / dashboards tab (which got removed) 
- [x] Remove `JackAndLevi` from the git history 😢 


<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #38718

leventebalogh committed:
feat(Plugins/Catalog): start adding necessary apis

leventebalogh committed:
feat(PLugins/Catalog): add extra helpers for merging local & remote plugins

leventebalogh committed:
feat(Plugins/Catalog): add plugin details as an optional field of CatalogPlugin

leventebalogh committed:
feat(PLugins/Catalog): add scaffolding for the new redux model

leventebalogh committed:
feat(PLugins/Catalog): export reducers based on a feature-flag

leventebalogh committed:
refactor(Plugins/Admin): rename api methods

leventebalogh committed:
feat(Plugin/Catalog): add an api method for fetching a single plugin

leventebalogh committed:
feat(Plugins/Admin): try cleaning stuff around plugin fetching

leventebalogh committed:
ffeat(Plugins/Catalog): return the catalog reducer when the feature flag is set

leventebalogh committed:
refactor(Plugins/Admin): fix typings

leventebalogh committed:
feat(Plugins/Admin): use the new reducer for the browse page

leventebalogh committed:
refactor(Plugins/Details): rename page prop type

jackw committed:
feat(catalog): introduce selectors to search and filter plugins list

leventebalogh committed:
refactor(Plugins/Admin): add a const for a state prefix

leventebalogh committed:
refactor(Plugins/Admin): use the state prefix in the actions

leventebalogh committed:
feat(Plugins/Admin): add types for the requests

leventebalogh committed:
refactor(Plugins/Admin): add request info to the reducer

leventebalogh committed:
refactor(Plugins/Admin): add request handling to the hooks & selectors

leventebalogh committed:
refactor(Plugins/Details): start using the data stored in Redux

leventebalogh committed:
refactor(Plugins/Admin): rename selector to start with "select"

leventebalogh committed:
fix(Plugins/Admin): only fetch plugins once

leventebalogh committed:
refactor(Plugins/Admin): make the tab selection work in details

jackw committed:
refactor(catalog): put back loading and error states in plugin list

leventebalogh committed:
refactor(Plugins/Admin): use CatalogPlugin for <PluginDetailsSignature />

leventebalogh committed:
feat(Plugins/Admin): add an api method for fetching plugin details

leventebalogh committed:
refactor(Plugins/Admin): add action for updating the details

leventebalogh committed:
irefactor(Plugins/Admin): show basic plugin details info

leventebalogh committed:
refactor(Plugin Details): migrate the plugin details header

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=38876) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=38876) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: leventebalogh<br/>:white_check_mark: jackw<br/>:x: Jack & Levi<br/><hr/>**Jack & Levi** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=38876) it.</sub>

leventebalogh committed:
refactor(Plugins/Admin): make the config and dashboards tabs work

leventebalogh committed:
refactor(Plugins/Admin): add old reducer state to the new one

jackw committed:
feat(catalog): introduce actions, reducers and hooks for install & uninstall

jackw committed:
refactor(catalog): wire up InstallControls component to redux

jackw committed:
refactor(catalog): move parentUrl inside PluginDetailsHeader and uncomment InstallControls

jackw committed:
feat(catalog): introduce code for plugin updates to install action

leventebalogh committed:
refactor(Plugins/Admin): add backward compatible actions

jackw committed:
test(catalog): update PluginDetails and Browse tests to work with catalog store

leventebalogh committed:
refactor(Plugins/Admin): make the dashboards and panels work again

leventebalogh committed:
refactor(Plugins/Admin): fix linter and typescript errors

leventebalogh committed:
fix(Plugins/Admin): put the local-only plugins to the beginning of the list

leventebalogh committed:
fix(Plugins/Admin): fix the mocks in the tests for PluginDetails

leventebalogh committed:
refactor(Plugins/Admin): remove unecessary hook usePluginsByFilter()

leventebalogh committed:
refactor(Plugins/Admin): extract the useTabs() hook to its own file

leventebalogh committed:
refactor(Plugins/Admin): remove unused helpers and types

leventebalogh committed:
fix(Plugins/Admin): show the first tab when uninstalling an app plugin

This can cause the user to find themselves on a dissappeared tab, as the
config and dashboards tabs are removed.

jackw committed:
fix(catalog): correct logic for checking if activeTabIndex is greater than total tabs

leventebalogh committed:
fix(Plugins/Admin): fix race-condition between fetching plugin details and all plugins

leventebalogh committed:
fix(Plugins): fix strict type errors

jackw committed:
chore(catalog): remove todos

jackw committed:
feat(catalog): render an alert in PluginDetails when a plugin cannot be found

jackw committed:
feat(catalog): use the proper store state

mckn commented on public/app/features/plugins/admin/api.ts:
```diff
@@ -1,20 +1,48 @@
+import { mergeLocalsAndRemotes, mergeLocalAndRemote } from './helpers';
+
+export async function getCatalogPlugins(): Promise<CatalogPlugin[]> {
+  const localPlugins = await getLocalPlugins();
+  const remotePlugins = await getRemotePlugins();
```
```suggestion
  const [localPlugins, remotePlugins] = await Promise.all([
    getLocalPlugins(),
    getRemotePlugins(),
  ]);
```

leventebalogh replied:
Nice catch! 👍

mckn commented on public/app/features/plugins/admin/components/InstallControls/InstallControlsButton.tsx:
```diff
@@ -1,68 +1,54 @@
   const styles = useStyles2(getStyles);
+  const uninstallBtnText = isUninstalling ? 'Uninstalling' : 'Uninstall';
 
   const onInstall = async () => {
```
Nit: could be a good idea to wrap this in a `useAyncFn` to prevent the risk of calling `setHasInstalledPanel` after the component has unmounted.

mckn replied:
The same thing goes for the other async functions in this component.

leventebalogh replied:
Yea, I think that's a fair point, will change 👍

leventebalogh committed:
refactor(Plugins/Admin): fetch local and remote plugins in parallell

Co-authored-by: Marcus Andersson <marcus.andersson@grafana.com>

mckn commented on public/app/features/plugins/admin/components/PluginDetailsBody.tsx:
```diff
@@ -1,65 +1,68 @@
         return (
           <div className={styles.container}>
-            <configPage.body plugin={plugin} query={{}} />
+            {/* TODO: we should pass the query params down */}
```
Something we should do within this PR or in a follow up PR?

leventebalogh replied:
I would personally fix this in an upcoming small PR if that's not a problem.

(Mostly just to not make this PR bigger and also because it is not something that was introduced now, but something that caught our eye and wanted to put a reminder on it.)

Is it ok if I create an issue for it?

mckn replied:
Yes, no worries! I just asked to make sure we didn't forgot about it :)

mckn commented on public/app/features/plugins/admin/pages/Browse.tsx:
```diff
@@ -29,8 +31,9 @@ export default function Browse({ route }: GrafanaRouteComponentProps): ReactElem
-  const sortedPlugins = plugins.sort(sorters[sortBy]);
+  const filteredPlugins = useSelector(find(q, filterBy, filterByType));
+  const { isLoading, error } = useFetchStatus();
+  const sortedPlugins = filteredPlugins.sort(sorters[sortBy]);
```
Would it make sense to introduce a `useGetAllAndFilter` hook to make it a little bit easier to follow flow of the information?

I get that the `useGetAll()` will trigger a async call to fetch the plugins. Then we use the `useFetchStatus()` to get the status of the current ongoing request. We use the `useSelector(find(...))` to get filtered plugins that we later sort.

So my question is: would it be easier to follow if we did something like the following?

```ts
const filter = find(q, filterBy, filterByType);
const sorter = sorters[sortBy];
const { isLoading, error, plugins } = useFetchPluginsAsync(filter, sorter);
```

mckn replied:
My thinking is that the `useGetAll` only seems to be used in this place so it can then be replaced by a more specific hook that reads a little bit easier.

leventebalogh replied:
I like those ideas, it could make it easier to read and reason about as well.

Although I am not against the new `useGetAllAndFilter()` hook, just bringing up another idea, what do you think about something like the following? Or do you find the separate new hook more straightforward?
```javascript
const { isLoading, error, plugins }  = useGetAll({ filters });
```

mckn replied:
I think that looks great! I just hoped that it would make the code easier to read and reason about. So I am super happy with your suggestion!

hugohaggmark commented on public/app/features/plugins/state/reducers.test.ts:
```diff
@@ -14,7 +15,7 @@ describe('pluginsReducer', () => {
     it('then state should be correct', () => {
       reducerTester<PluginsState>()
-        .givenReducer(pluginsReducer, { ...initialState })
+        .givenReducer(pluginsReducer as Reducer<PluginsState, AnyAction>, { ...initialState })
```
Why all these `as Reducer<PluginsState, AnyAction>` declarations, something that can be improved in `reducerTester`?

leventebalogh replied:
Very fair question @hugohaggmark, I think it at least deserves an explanatory comment.

I think the `reducerTester` is actually fine, why we needed it was because now we are exporting the `pluginsReducer` conditionally based on a feature flag (if the feature flag is enabled we use the "new" reducer, which has two extra fields, while by default we are loading the "old" one).

Due to this logic TS sees it as a union type which is not fulfilled by the `pluginsReducer` that is used in these "old" tests.

**We manually set the types here because this test is only testing the functionality of the "old" reducer, and it cannot really happen that the old reducer meets "new" reducer state containing the extra fields.**

Sorry for the lengthy explanation, please let me know if something is not clear.

It can easily be though that we didn't take the best approach!

hugohaggmark replied:
Great explanation! I'm always interested in if the `reducerTester` needs to be extended more. Thank you!

mckn reviewed (approved):
Wow, great work on this!!

![hat off for you](https://media.giphy.com/media/JNm39AP1Wx10zFV3N9/giphy.gif?cid=ecf05e477v9dh387x9az0ctztttvxd41tzcr68sdll60kz1m&rid=giphy.gif&ct=g)

jackw committed:
style(catalog): fix prettier error in api

jackw committed:
fix(catalog): prevent throwing error if InstallControlsButton is unmounted during install

leventebalogh committed:
refactor(Plugins/Admin): add a separate hook for filtering & sorting plugins
